basics/section_2-46.rb

```diff
@@ -0,0 +1,58 @@
+require "bcrypt"
+
+my_password = BCrypt::Password.create("my password")
+my_password_1 = BCrypt::Password.create("my password")
+my_password_2 = BCrypt::Password.create("my password")
+
+puts my_password
+puts my_password_1
+puts my_password_2
+
+# When we are checking with this operation with a hashed pass, what actually
+# happens is a hash verification and not a straight string comparison.
+
+puts my_password == my_password_2 # false
+puts my_password == "my password" # true
+puts "my password" == my_password # false
+
+# The equality operator == (the compare function) is a method on an object.
+
+# In the case of a String object being to the left, it is the equality operator
+# method of the String that gets called.
+# In the case of the hash digest object being on the left, it's equality operator is called.
+
+#######################################################
+
+users = [
+  { username: "sean", password: "password1" },
+  { username: "test", password: "password2" },
+  { username: "demo", password: "password3" },
+  { username: "account", password: "password4" },
+  { username: "ho", password: "password5" }
+]
+
+def create_hash_digest(password)
+  BCrypt::Password.create(password)
+end
+
+def verify_hash_digest(password)
+  BCrypt::Password.new(password)
+end
+
+def create_secure_users(list_of_users)
+  list_of_users.each do |user_record|
+    user_record[:password] = create_hash_digest(user_record[:password])
+  end
+  return list_of_users
+end
```

Comment:
```suggestion
def create_secure_users(users)
  users.map{|user|
    user[:password] = create_hash_digest(user[:password])
    user
  }
end
```

Reply:
Is `map` preferred here as it doesn't mutate the original array?

Reply:
Preferred because the goal here is transformation of the data within the array.